src/main/java/com/google/api/gax/bundling/ThresholdBundler.java

```diff
@@ -129,11 +129,14 @@ private Builder() {
    *
    * @throws FlowControlException
    */
-  public void add(E e) throws FlowControlException {
+  public synchronized void add(E e) throws FlowControlException {
     final Lock lock = this.lock;
+    flowController.reserve(e);
+    // We need to reserve resources from flowController outside the lock, but we also need to
+    // prevent concurrent calls to add(E e) from all reserving flowController resources and then
+    // waiting to acquire the lock, so this method is marked as synchronized.
```

Comment:
Because otherwise in the case of LimitExceededBehavior.Block we will cause a deadlock. This thread holds the lock and blocks on flowController.reserve(e), while the takeBundle() thread cannot acquire the lock, and thus cannot call flowController.release(e) to unblock this thread.

If we do not want to use `synchronized` then I will add another internal lock.